### DISCUSSION TITLE
How does setuptools determine the appropriate site-packages directory?

### DISCUSSION THREAD

**Discussion by adamjstewart** [pypa/setuptools#2813] [General]
Hi, I'm a developer for the [Spack](https://spack.io) package manager. Spack is similar to Conda in that we have both Python and non-Python packages, and similar to Nix in that we install all packages to a separate prefix allowing for multiple versions of the same software to be installed.

When dealing with Python installations, we've noticed that third-party Python libraries may be installed in one of several directories:

* `lib/pythonX.Y/site-packages` on most systems
* `lib64/pythonX.Y/site-packages` on RHEL/CentOS/Fedora
* `lib/pythonX/dist-packages` on Debian/Ubuntu

We would like to know this directory ahead of time so that we can set `PYTHONPATH` appropriately and ensure that Python libraries are installed to the same prefix as the system Python expects. Right now, we're using distutils to determine this:
```python
from distutils.sysconfig import get_python_inc, get_python_lib

for plat_specific in [True, False]:
    print(get_python_inc(plat_specific, prefix=""))
    for standard_lib in [True, False]:
        print(get_python_lib(plat_specific, standard_lib, prefix=""))
```
We then use these values to explicitly tell setuptools where to install the package to:
```console
$ python setup.py install --root=<prefix> --install-purelib=<not-plat-specific-not-standard-lib> --install-platlib=<plat-specific-not-standard-lib> --install-scripts=bin --install-data= --install-headers=<plat-specific>
```
This seems to work most of the time, but I've noticed that some packages like cython will install to the `--install-platlib` directory while others like setuptools will install to the `--install-purelib` directory. Also, distutils is deprecated in Python 3.10 and will be removed in 3.12, so we would like to move away from this approach. Also also, we don't want to have to run Python if possible since we may be cross-compiling. We could simply recursively search for a particular prefix, but we need to know before installation where libraries will get installed.

My questions are as follows:

1. If I run `python setup.py install`, how does setuptools determine the appropriate site-packages directory?
2. For ancient packages that still use distutils, is this behavior the same? If not, how does distutils handle things?
3. If we want to ensure that installation prefixes follow the same structure as the main Python installation, do we need to specify `--install-*` like above?

Pinging other Spack devs who would love to know the answers to how best to package Python libraries: @tgamblin @alalazo @skosukhin @permeakra @wspear @robertu94 @dev-zero 

Links to Spack issues and PRs for additional context: https://github.com/spack/spack/pull/24095 https://github.com/spack/spack/pull/21446 https://github.com/spack/spack/issues/24076 https://github.com/spack/spack/issues/24526 https://github.com/spack/spack/pull/25998 https://github.com/spack/spack/issues/26546

**Comment by jwodder** [comment]
FYI, you should be asking about this question of [pip](https://github.com/pypa/pip), not setuptools.  Setuptools' installation capabilities are deprecated and discouraged nowadays in favor of pip.

**Reply by adamjstewart** [reply]
Ah, I didn't realize that. We've been trying to avoid installing an entire package manager (pip) just to install Python libraries with our own package manager (Spack) and setuptools' installation capabilities were very handy for this. Is there some documentation somewhere saying that this is deprecated? When is it planned to be removed?

I've also noticed a few other alternate build systems like poetry/flit that are becoming popular and those also seem to rely on pip for installation. It would probably be overkill to implement our own PEP-XYZ-compliant installation protocol since we're a general purpose package manager, so we may just have to start using pip for everything...

**Reply by tgamblin** [reply]
As long as we can avoid using wheels, I think this approach would work for us.  We need to be able to produce our own "binaries" from the source.

**Reply by webknjaz** [reply]
https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html

**Reply by tgamblin** [reply]
Ok, so spack could also just adhere to the standards when executing `setup.py`.

**Reply by adamjstewart** [reply]
I think the point is that we shouldn't be executing `setup.py` at all. From what I can tell we should replace all invocations of:
```console
$ python setup.py build
$ python setup.py install
```
with:
```console
$ pip install
```
The trick then is figuring out the appropriate flags, and understanding where `pip` will decide to install the libraries (which site-packages dir). This will also solve a lot of other problems we've had with `PythonPackage`, including:

* no support for wheels
* no support for flit/poetry
* no support for setuptools when a `setup.py` file doesn't exist

I'll have to play around with this, but I think the right thing to do is just delegate to `pip`. If we can figure out a way to prevent `pip` from downloading missing dependencies that would be even better.

**Reply by jwodder** [reply]
@adamjstewart 

> If we can figure out a way to prevent `pip` from downloading missing dependencies that would be even better.

I think you may be after the [`--no-deps` option to `pip install`](https://pip.pypa.io/en/stable/cli/pip_install/#install-no-deps).

**Reply by adamjstewart** [reply]
Thanks, I'll have to see if that flag skips downloading missing deps and crashes or just skips checking for those deps entirely.

**Reply by tgamblin** [reply]
@adamjstewart: If `pip` is the right frontend to use for PEP517 and PEP518 then we should probably be using it.  I am not sure this is going to work for all builds, or if we can use the options we need.  We may need to use the lower-level pep517/pep518 modules or something else to do what we need.

In particular we use `--single-version-externally-managed` with `setuptools` right now, and we *don't* want to install from wheels.  If we can do the equivalents of those things with `pip` then I think we're good.

**Reply by adamjstewart** [reply]
> FYI, you should be asking about this question of pip, not setuptools.

Opened the same question [here](https://discuss.python.org/t/understanding-site-packages-directories/12959) to hopefully get responses from the broader Python packaging community.

**Reply by CAM-Gerlach** [reply]
Just FYI, if pip is not a good fit and you need a lower-level installation mechanism for modern PEP 517 packages designed for this use case (use by distros and package managers that control download and dep installation themselves), [installer](https://github.com/pradyunsg/installer) was created to address this use case and is the de-facto "blessed" solution (though it is still relatively early in development).

**Reply by adamjstewart** [reply]
@CAM-Gerlach I would actually love to use build/installer in our Python build system instead of pip. Spack is already a package manager, and we go out of our way to disable all of pip's package management features just because we need something to build and install a single wheel.

Unfortunately, build/installer currently require build/installer to build and install. Until build/installer vendor all of their dependencies like pip/setuptools/flit-core/etc. do, it isn't possible to install build/installer without an existing installation of build/installer. We could use pip to install build/installer, but at that point why not just use pip for everything?

**Reply by CAM-Gerlach** [reply]
Well, installer is pure Python and doesn't require any dependencies, so it has no need of vendoring them. You can simply extract the wheel (its just a zip) and run installer from there (either to install itself, or other packages). `build` only needs two required deps, `packaging` and `pep517` (three on Python <=3.7, including `importlib_metadata`), so its not that tough to bootstrap, and that should only be required when *building* spack packages, not installing them (which only needs `installer`). There are a number of past discussions on this on the Python discourse; feel free to ask there for more.

**Reply by adamjstewart** [reply]
Thanks, it looks like the latest stable release of build requires [5 run-time dependencies](https://github.com/pypa/build/blob/0.7.0/setup.cfg#L28,L32) while the latest stable release of installer requires [2 run-time dependencies](https://github.com/pradyunsg/installer/blob/0.4.0/pyproject.toml#L14,L15). The development version of installer has dropped these dependencies, but also dropped support for older versions of Python.

Unfortunately, Spack still supports installations with Python 2, as many scientific software libraries have been lax on porting. I'm working on deprecating this in https://github.com/spack/spack/pull/28003 but it will be another full release cycle before we can remove this entirely.

Once Spack only supports Python 3.7+ installations, I'll definitely be eager to try to port our build system to build/installer. Even if build still has run-time dependencies, I can install them from wheels with installer (assuming they are pure-Python) in order to bootstrap build.

**Reply by CAM-Gerlach** [reply]
> Thanks, it looks like the latest stable release of build requires 5 run-time dependencies while the latest stable release of installer requires 2 run-time dependencies. The development version of installer has dropped these dependencies, but also dropped support for older versions of Python.

The deps are still only needed on those older versions of Python (note the conditional in the spec). If you're running on a modern, supported version of Python, they will not be installed even in that version of `installer`.

> Unfortunately, Spack still supports installations with Python 2, as many scientific software libraries have been lax on porting. I'm working on deprecating this in spack/spack#28003 but it will be another full release cycle before we can remove this entirely.

Is that so? Based on my experience as a scientific software developer myself, the scientific Python ecosystem led the way in not only porting to Python 3, but also dropping Python 2. Its been a long time since I've seen a significant scientific Python project that was Python 2 only, and almost all the heavy hitters have dropped Python 2 support by now.

Realistically, any project that hasn't ported to the decade+ old Python 3 likely hasn't even considered moving beyond a legacy Setuptools (or even, _shudder_, `distutils` based build system), so the intersection of the two sets is likely zero, and you are thus unlikely to need `installer` for projects that don't support it with no dependencies.

> Once Spack only supports Python 3.7+ installations, I'll definitely be eager to try to port our build system to build/installer. Even if build still has run-time dependencies, I can install them from wheels with installer (assuming they are pure-Python) in order to bootstrap build.

Keep in mind that increasing number of projects (including many of the core PyPA ones) rely on other non-Setuptools-based build systems and wheel-only distributrion, particularly with `flit` being targeted at scientific libraries, `scikit-build` aiming for heavy-duty PyData and machine learning packages, and Numpy, Scipy and others switching to Meson, and thus you are going to run into an increasing number of projects, including critical ones, that are uninstallable with deprecated legacy build tooling. It is your choice to support end of life, obsolete Python versions, but I just want to make sure you're fully aware of the growing consequences of doing so for you and your users.

**Reply by adamjstewart** [reply]
> Based on my experience as a scientific software developer myself, the scientific Python ecosystem led the way in not only porting to Python 3, but also dropping Python 2. Its been a long time since I've seen a significant scientific Python project that was Python 2 only, and almost all the heavy hitters have dropped Python 2 support by now.

That's certainly true for things like numpy/scipy/matplotlib, but is definitely not the case for specialty research code, much of which doesn't have the funding to be maintained after it was first released. See [here](https://github.com/spack/spack/pull/28003#issuecomment-1010404640) for some examples.

> considered moving beyond a legacy Setuptools

Is setuptools considered legacy? My impression was the setuptools was still the recommended build backend, although flit/poetry are also supported.

> Keep in mind that increasing number of projects (including many of the core PyPA ones) rely on other non-Setuptools-based build systems and wheel-only distributrion, particularly with flit being targeted at scientific libraries, scikit-build aiming for heavy-duty PyData and machine learning packages, and Numpy, Scipy and others switching to Meson, and thus you are going to run into an increasing number of projects, including critical ones, that are uninstallable with deprecated legacy build tooling. It is your choice to support end of life, obsolete Python versions, but I just want to make sure you're fully aware of the growing consequences of doing so for you and your users.

Yes, I'm aware of the consequences. Spack allows you to have multiple versions of the same library, so concretizing and ensuring that versions of the library are compatible is actually really easy. I'm still trying to move Spack to only support non-EOL versions of Python (https://github.com/spack/spack/pull/28003) but there is significant push back.

**Reply by abravalheri** [reply]
> Is setuptools considered legacy? My impression was the setuptools was still the recommended build backend, although flit/poetry are also supported.

Hi @adamjstewart, `setuptools` is not considered legacy 😄 (or even deprecated). The project is pretty much alive.

What is considered deprecated is calling `python setup.py` directly *and/or* not having a `pyproject.toml` file with PEP 518 information (also there is a `setuptools.build_meta:legacy` that is only there as "last ressort").

**Reply by adamjstewart** [reply]
> What is considered deprecated is calling `python setup.py` directly

Yep, we are moving to `pip install` (and hopefully someday build/installer) in https://github.com/spack/spack/pull/27798. This will support all versions of Python that pip has ever supported and work even for packages with no `setup.py` or that don't use setuptools.

**Reply by CAM-Gerlach** [reply]
> That's certainly true for things like numpy/scipy/matplotlib, but is definitely not the case for specialty research code, much of which doesn't have the funding to be maintained after it was first released. See here for some examples.

I see; I've always thought it was rather unfortunate that the funding model is the way it is, and have advocated for change on that front, but I also sympathize with those stuck with it. For an free and open source, volunteer project, you have an equal right to spend your time supporting EoL Python versions as you do dropping them, and I'm glad you've taken the needs of your community into account here.

> Is setuptools considered legacy? My impression was the setuptools was still the recommended build backend, although flit/poetry are also supported.

Sorry, I can see why my wording was confusing especially since I ended the parenthetical in the wrong place—I mean to say "legacy-Setuptools- (or, _shudder_, distutils-) based build system", not "legacy Setuptools (or, _shudder_, distutils based build system)", referring to the legacy (`setup.py`-based) Setuptools backend (`setuptools.build_meta:__legacy__`), rather than the modern PEP 517 (setup.cfg/pyproject.toml-based) one, not Setuptools itself being legacy (though long-term in the decade+ timespan, it seems to be the expectation that the community will eventually move on to one or more of the more modern build systems without all of Setuptools' remaining warts, though that's purely speculative and will be up to package authors to decide).

Also, to be clear, Setuptools is no more officially/generally "recommended" than are Flit, Poetry, Pipenv, PDM or any other standards-conforming packaging tool; each has specific use cases they are appropriate for (e.g. Flit-pure Python libraries, Poetry-heavy duty libraries and applications, Pipenv- applications, Scikit-Build- compiled scientific libraries, etc).

> there is a `setuptools.build_meta:legacy` [sic] that is only there as "last ressort" [sic]

Yeah, this is what I was referring to, and directly calling setup.py/not using `pyproject.toml` more generally as you also mention; it was quite confusing due to how I structure the syntax, sorry.
